src/server/game/AI/SmartScripts/SmartAI.cpp

```diff
@@ -650,9 +650,7 @@ void SmartAI::OnCharmed(bool apply)
 {
     GetScript()->ProcessEventsFor(SMART_EVENT_CHARMED, NULL, 0, 0, apply);
 
-    if (!apply && !me->IsInEvadeMode() && me->GetCharmerGUID())
-        if (Unit* charmer = ObjectAccessor::GetUnit(*me, me->GetCharmerGUID()))
-            AttackStart(charmer);
+    CreatureAI::OnCharmed(apply);
```

Comment:
I totally agree with u, SAI Charm mechanics are incomplete.

My "solution" works. There are several important points:
* If its one of the cases in which no PetAI is needed, just store the ReactState on charm and restore on remove. On a clean TC it works, I've tested it.
* SmartScripts on charms without PetAI continue and events like casts or texts do trigger. Example: http://www.wowhead.com/item=38302/ruby-beacon-of-the-dragon-queen
* SmartScripts on charms which replace SAI for PetAI dont trigger. Like I said this is the incorrect point of this solution, to avoid handling PetAI inside SAI.

The perfect solution is still the last one I wrote down, implement PetAI inside SAI.
Like I said, huge effort for a small group of cases; but still the right thing to do.


To handle events that should not be triggered while charm, for example, we could implement a flag like ````SMART_EVENT_FLAG_TRIGGER_ONCHARM = 512````